### PR TITLE
add set/get volume (reference nao_audio package)

### PR DESCRIPTION
**Pull request by kochigami**
added services for getting & setting master volume.
This is a new version of https://github.com/ros-naoqi/naoqi_driver/pull/89

reference: http://doc.aldebaran.com/2-5/naoqi/audio/alaudiodevice-api.html#alaudiodevice-api
```ALAudioDeviceProxy::setOutputVolume``` and ```ALAudioDeviceProxy::getOutputVolume```

I added the condition of ```master_volume < 0``` or ```master_volume > 100``` so that it is in accordance with ```nao_audio``` in ```nao_interaction```.
I used srv files in ```nao_interaction_msgs``` in ```nao_interaction```.

reference:
I sent https://github.com/ros-naoqi/nao_interaction/pull/10

### CODE REVIEW

**Comment by kochigami** [comment]
For the time being, I didn't know how to solve the error below when I did ```catkin bt```.
This can't be merged now.
```
Errors     << naoqi_driver:make /home/kochigami/catkin_ws/logs/naoqi_driver/build.make.136.log
/home/kochigami/catkin_ws/devel/.private/naoqi_driver/lib/libnaoqi_driver.so: undefined reference to `naoqi::service::GetVolumeService::reset(ros::NodeHandle&)'
/home/kochigami/catkin_ws/devel/.private/naoqi_driver/lib/libnaoqi_driver.so: undefined reference to `naoqi::service::SetVolumeService::reset(ros::NodeHandle&)'
/home/kochigami/catkin_ws/devel/.private/naoqi_driver/lib/libnaoqi_driver.so: undefined reference to `naoqi::service::GetVolumeService::GetVolumeService(std::string const&, std::string const&, boost::shared_ptr<qi::Session> const&)'
/home/kochigami/catkin_ws/devel/.private/naoqi_driver/lib/libnaoqi_driver.so: undefined reference to `naoqi::service::SetVolumeService::SetVolumeService(std::string const&, std::string const&, boost::shared_ptr<qi::Session> const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [/home/kochigami/catkin_ws/devel/.private/naoqi_driver/lib/naoqi_driver/naoqi_driver_node] Error 1
make[1]: *** [CMakeFiles/naoqi_driver_node.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by kochigami** [comment]
I found the above error was caused by the wrong position of CMakeLists.txt.
I just copied modified files from https://github.com/ros-naoqi/naoqi_driver/pull/89 , and then I made a mistake when relocating them.
Sorry about that.

I confirmed this pull-request under ubuntu 14.04, ROS indigo, Pepper with NAOqi 2.5.5.5.

```roslaunch naoqi_driver naoqi_driver.launch network_interface:=eth1```

```
kochigami@kochigami-ThinkPad-T450:~/catkin_ws/src/naoqi_driver/share$ rosservice call /naoqi_driver/get_master_volume "{}" 
master_volume: 
  data: 30
kochigami@kochigami-ThinkPad-T450:~/catkin_ws/src/naoqi_driver/share$ rosservice call /naoqi_driver/set_master_volume "master_volume:
  data: -100" 

kochigami@kochigami-ThinkPad-T450:~/catkin_ws/src/naoqi_driver/share$ rosservice call /naoqi_driver/get_master_volume "{}" 
master_volume: 
  data: 30
kochigami@kochigami-ThinkPad-T450:~/catkin_ws/src/naoqi_driver/share$ rosservice call /naoqi_driver/set_master_volume "master_volume:
  data: 10" 

kochigami@kochigami-ThinkPad-T450:~/catkin_ws/src/naoqi_driver/share$ rosservice call /naoqi_driver/get_master_volume "{}" 
master_volume: 
  data: 10
kochigami@kochigami-ThinkPad-T450:~/catkin_ws/src/naoqi_driver/share$ rosservice call /naoqi_driver/set_master_volume "master_volume:
  data: 200" 

kochigami@kochigami-ThinkPad-T450:~/catkin_ws/src/naoqi_driver/share$ rosservice call /naoqi_driver/get_master_volume "{}" 
master_volume: 
  data: 10
```

**Comment by nlyubova** [comment]
Thank you @kochigami ! Do you have a possibility to check it with  ROS Kinetic? and with Nao robot?

**Comment by kochigami** [comment]
I'm sorry for my late reply.

I've tested with Pepper + ROS kinetic.
I haven't tested with NAO + ROS indigo/ kinetic yet. Please wait for a while.

What I did is as follows:
- fetching https://github.com/ros-naoqi/nao_interaction/pull/10 & ```catkin bt```
- roscd naoqi_driver & catkin bt
- I found these errors below:

```
Errors     << naoqi_driver:make /home/pepper/catkin_ws/logs/naoqi_driver/build.make.000.log                                                                                                               
In file included from /home/pepper/catkin_ws/src/naoqi_driver/src/helpers/driver_helpers.cpp:18:0:
/home/pepper/catkin_ws/src/naoqi_driver/src/helpers/driver_helpers.hpp:26:55: fatal error: nao_interaction_msgs/SetAudioMasterVolume.h: No such file or directory
compilation terminated.
make[2]: *** [CMakeFiles/naoqi_driver.dir/src/helpers/driver_helpers.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /home/pepper/catkin_ws/src/naoqi_driver/include/../src/converters/converter_base.hpp:25:0,
                 from /home/pepper/catkin_ws/src/naoqi_driver/include/../src/converters/audio.hpp:24,
                 from /home/pepper/catkin_ws/src/naoqi_driver/src/event/audio.hpp:37,
                 from /home/pepper/catkin_ws/src/naoqi_driver/src/event/audio.cpp:30:
/home/pepper/catkin_ws/src/naoqi_driver/include/../src/converters/../helpers/driver_helpers.hpp:26:55: fatal error: nao_interaction_msgs/SetAudioMasterVolume.h: No such file or directory
compilation terminated.
make[2]: *** [CMakeFiles/naoqi_driver.dir/src/event/audio.cpp.o] Error 1
In file included from /home/pepper/catkin_ws/src/naoqi_driver/src/converters/converter_base.hpp:25:0,
                 from /home/pepper/catkin_ws/src/naoqi_driver/src/converters/audio.hpp:24,
                 from /home/pepper/catkin_ws/src/naoqi_driver/src/naoqi_driver.cpp:27:
/home/pepper/catkin_ws/src/naoqi_driver/src/converters/../helpers/driver_helpers.hpp:26:55: fatal error: nao_interaction_msgs/SetAudioMasterVolume.h: No such file or directory
compilation terminated.
make[2]: *** [CMakeFiles/naoqi_driver.dir/src/naoqi_driver.cpp.o] Error 1
make[1]: *** [CMakeFiles/naoqi_driver.dir/all] Error 2
make: *** [all] Error 2
```
- when I added https://github.com/ros-naoqi/naoqi_driver/pull/97/commits/45c885f3d96316d4f18acb3be136d2e4cf0fff10 ,
I could compile naoqi_driver and confirmed ```rosservice call /naoqi_driver/get_master_volume "{}" ``` and ```rosservice call /naoqi_driver/set_master_volume "master_volume:
  data: -100, 30"  ``` worked fine.
- I'm not sure if this change is enough or not due to my lack of study. I will study how to compile ```naoqi_driver``` package.

**Comment by kochigami** [comment]
I also confirmed this pull-request with NAO + ROS indigo, kinetic.

I learned that naoqi_driver package is compiled by using qibuild or catkin.
CMakeLists.txt is for qibuild, and CMakeLists_catkin.txt is for catkin.
That is why we need to modify CMakeLists.txt and CMakeLists_catkin.txt.
 
After https://github.com/ros-naoqi/nao_interaction/pull/10 is merged, 
I think this is ok to be merged.

**Comment by nlyubova** [comment]
Thank you for testing ! 
Right, that there are two CMakeLists for qibuild and catkin and sorry for not making it clear before ...
I've just merged  ros-naoqi/nao_interaction#10

**Comment by kochigami** [comment]
Thank you for your comment.
Next, I will learn how to pass the Travis test because I'm a novice at it :)

Travis error says:
```
In file included from /catkin_ws/src/naoqi_driver/src/helpers/driver_helpers.cpp:18:0:
/catkin_ws/src/naoqi_driver/src/helpers/driver_helpers.hpp:26:55: fatal error: nao_interaction_msgs/SetAudioMasterVolume.h: No such file or directory
 #include <nao_interaction_msgs/SetAudioMasterVolume.h>
                                                       ^
compilation terminated.
In file included from /catkin_ws/src/naoqi_driver/src/converters/converter_base.hpp:25:0,
                 from /catkin_ws/src/naoqi_driver/src/converters/audio.hpp:24,
                 from /catkin_ws/src/naoqi_driver/src/naoqi_driver.cpp:27:
/catkin_ws/src/naoqi_driver/src/converters/../helpers/driver_helpers.hpp:26:55: fatal error: nao_interaction_msgs/SetAudioMasterVolume.h: No such file or directory
 #include <nao_interaction_msgs/SetAudioMasterVolume.h>
                                                       ^
compilation terminated.
```

**Comment by kochigami** [comment]
When I looked into #87, I found I also made a mistake on this pull-request.
This somehow works on Pepper, NAO, ros indigo, kinetic though...

I will update this later:
my working branch is here:
https://github.com/kochigami/naoqi_driver/tree/add-set-get-volume-fix

**Comment by kochigami** [comment]
I modified function types of `get/set Volume` like #87.
I confirmed with:
- [x] Pepper (NAOqi 2.5.5.5) + ros indigo 
- [x] Pepper (NAOqi 2.5.5.5) + ros kinetic 
- [x] NAO (NAOqi 2.1.4) + ros indigo
- [x] NAO (NAOqi 2.1.4) + ros kinetic

File: share/boot_config.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
-/* QQVGA = 0, QVGA = 1, VGA =2 */
 {
+  "_comment": "QQVGA = 0, QVGA = 1, VGA = 2",
```
Comment (nlyubova) [inline share/boot_config.json]: Should be in a different Pull Request because it is not related to the goal of this PR

**Comment by nlyubova** [comment]
could you do "git pull --rebase" do get last changes ? and then commit?

File: share/boot_config.json (hunk 1 above)
Comment (nlyubova) [reply share/boot_config.json]: Ah, I see. Just remove this commit from your Pull Request

Comment (kochigami) [reply share/boot_config.json]: Thank you very much for your review.
I removed this commit.

File: src/services/set_volume.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  void SetVolumeService::reset( ros::NodeHandle& nh )
+  {
+    service_ = nh.advertiseService(topic_, &SetVolumeService::callback, this);
+  }
+
+  bool SetVolumeService::callback( nao_interaction_msgs::SetAudioMasterVolumeRequest& req, nao_interaction_msgs::SetAudioMasterVolumeResponse& resp )
+  {
+    if (req.master_volume.data < 0 or req.master_volume.data > 100){
+      return true;
```
[the remark this answers is not in the document]
Comment (kochigami) [reply src/services/set_volume.cpp]: Thank you very much for your review.
I modified it.

If we return false, the output becomes as follows:
```
rosservice call /naoqi_driver/set_master_volume "master_volume:
  data: 1000" 
ERROR: service [/naoqi_driver/set_master_volume] responded with an error:
```
I also add `ROS_ERROR` message.

File: src/helpers/driver_helpers.cpp [hunk 3]
```diff
@@ -194,6 +194,57 @@ const naoqi_bridge_msgs::RobotInfo& getRobotInfo( const qi::SessionPtr& session
...
+
+const int& getVolume( const qi::SessionPtr& session, nao_interaction_msgs::GetAudioMasterVolumeRequest req)
+{
+  static int volume;
+  try{
+    volume = getVolumeLocal(session, req);
+    return volume;
+  }
+  catch(const std::exception& e){
+    return volume;
```
[the remark this answers is not in the document]
Comment (kochigami) [reply src/helpers/driver_helpers.cpp]: Thank you very much for your review.
I modified it to return volume = -1 in catch.

File: src/services/get_volume.cpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  void GetVolumeService::reset( ros::NodeHandle& nh )
+  {
+    service_ = nh.advertiseService(topic_, &GetVolumeService::callback, this);
+  }
+
+  bool GetVolumeService::callback( nao_interaction_msgs::GetAudioMasterVolumeRequest& req, nao_interaction_msgs::GetAudioMasterVolumeResponse& resp )
+  {
+    resp.master_volume.data = helpers::driver::getVolume(session_, req);
+    return true;
```
[the remark this answers is not in the document]
Comment (kochigami) [reply src/services/get_volume.cpp]: Thank you very much for your review.
I modified it to check returned value is -1 or not. 
When returned value is -1, the output becomes as follows:
```
rosservice call /naoqi_driver/get_master_volume "{}" 
ERROR: service [/naoqi_driver/get_master_volume] responded with an error: 
```

**Comment by kochigami** [comment]
Thank you very much for your review.
I updated the pull request and confirmed with
- [x] Pepper (NAOqi 2.5.5.5) + ros indigo 
- [ ] Pepper (NAOqi 2.5.5.5) + ros kinetic 
- [ ] NAO (NAOqi 2.1.4) + ros indigo
- [ ] NAO (NAOqi 2.1.4) + ros kinetic

In addition, when doing `git rebase`, I made several commits which are not necessary (modify-functio-type and modify-try-catch).
After this pull request is approved, I will squash them. Sorry for my trouble.

**Comment by nlyubova** [comment]
@kochigami thanks for modifications and no worries about squashing, GitHub can squash when merging :)

**Comment by kochigami** [comment]
Thank you for your comment.
I confirmed this with four conditions above (Pepper/ NAO, indigo/kinetic), but I got an error at Travis.
The error says:
```
In file included from /catkin_ws/src/naoqi_driver/src/helpers/driver_helpers.cpp:18:0:
/catkin_ws/src/naoqi_driver/src/helpers/driver_helpers.hpp:28:41: fatal error: naoqi_bridge_msgs/SetString.h: No such file or directory
 #include <naoqi_bridge_msgs/SetString.h>
                                         ^
compilation terminated.
```
Maybe releasing `naoqi_bridge_msgs` is required. I will study it later.

**Comment by nlyubova** [comment]
Right, I seen this error on Travis and it is related to the last merged pull request. However, we've merged your corresponding PR in naoqi_bridhe_msgs so it should be able to find that file. We also have required naoqi_bridge_msgs dependence in package.xml and CMakeList that is good. I'll check it more precisely, and please let us know if we forgot something

**Comment by kochigami** [comment]
I learned that new deb file of `ros-indigo-naoqi-bridge-msgs` is released after about two weeks. 
In travis, the old version is used, which causes an error. 
I saw these:
http://wiki.ros.org/ShadowRepository
http://packages.ros.org/ros/ubuntu/pool/main/r/ros-indigo-naoqi-bridge-msgs/
http://packages.ros.org/ros-shadow-fixed/ubuntu/pool/main/r/ros-indigo-naoqi-bridge-msgs/

After waiting for two weeks, I will try to run travis.
(I'm sorry to try open/close this again and again.)

**Comment by nlyubova** [comment]
Right, it takes some time. I've released it yesterday,  no worries about this Travis error

**Comment by nlyubova** [comment]
It is already stable on buildfarm, so no worries.

**Comment by kochigami** [comment]
Now deb version of `naoqi-bridge-msgs` is 0.0.8.
I ran Travis again and found the newest version of `nao-interaction-msgs` is also needed.
Sorry about that.

travis log:
```
In file included from /catkin_ws/src/naoqi_driver/src/helpers/driver_helpers.cpp:18:0:
/catkin_ws/src/naoqi_driver/src/helpers/driver_helpers.hpp:30:55: fatal error: nao_interaction_msgs/SetAudioMasterVolume.h: No such file or directory
 #include <nao_interaction_msgs/SetAudioMasterVolume.h>
                                                       ^
compilation terminated.
```

**Comment by kochigami** [comment]
Sorry, I deleted the commits accidentally...
I just wanted to delete unnecessary commit to resolve conflicts, but I deleted other commits...
I will re-make this PR. I'm very sorry for my mistake.
